sdk/eidos_sdk/cpu/conversational_agent_cpu.py

```diff
@@ -59,56 +58,113 @@ async def schedule_request(
+            tool_result = await tool_def.execute(call_context=parent_stream_context, args=tc.arguments)
+            message = ToolResponseMessage(
+                logic_unit_name=tool_def.logic_unit.__class__.__name__,
+                tool_call_id=tc.tool_call_id,
+                result=tool_result,
+                name=tc.name,
+            )
+            await self.memory_unit.storeMessages(call_context, [message])
+            yield ObjectOutputEvent(stream_context=tool_context, content=tool_result)
+            yield EndStreamEvent(stream_context=tool_context, stop_reason=StopReason.COMPLETED)
+        except Exception as e:
+            yield ErrorEvent(stream_context=tool_context, reason=str(e))
 
     async def _llm_execution_cycle(
             self,
             call_context: CallContext,
-            conversation: List[LLMMessage],
             output_format: Union[Literal["str"], Dict[str, Any]],
-    ) -> AssistantMessage:
+    ) -> AsyncIterator[StreamEvent]:
         num_iterations = 0
         while num_iterations < self.spec.max_num_function_calls:
+            conversation = await self.memory_unit.getConversationHistory(call_context)
             tool_defs = await LLMToolWrapper.from_logic_units(self.logic_units, conversation=conversation)
-            assistant_message = await self.llm_unit.execute_llm(
-                call_context, conversation, [w.llm_message for w in tool_defs.values()], output_format
-            )
-            await self.memory_unit.storeMessages(call_context, [assistant_message])
-            if assistant_message.tool_calls:
-                calls = []
-                for tool_call in assistant_message.tool_calls:
-                    tool_def = tool_defs[tool_call.name]
-
-                    async def do_call(tc: ToolCall):
-                        tool_result = await tool_def.execute(call_context=call_context, args=tc.arguments)
-                        # todo, store tool response result as Any (must be json serializable) so that it can be retrieved symmetrically
-                        message = ToolResponseMessage(
-                            logic_unit_name=tool_def.logic_unit.__class__.__name__,
-                            tool_call_id=tc.tool_call_id,
-                            result=tool_result,
-                            name=tc.name,
+            tool_call_events = []
+            end_stream_event = None
+            execute_llm_ = self.llm_unit.execute_llm2(call_context, conversation, [w.llm_message for w in tool_defs.values()], output_format)
+            async for event in execute_llm_:
+                if isinstance(event, ToolCallEvent):
+                    tool_call_events.append(event)
+                elif isinstance(event, StartStreamEvent):
+                    yield event
+                elif isinstance(event, StringOutputEvent) or isinstance(event, ObjectOutputEvent):
+                    # todo -- change memory unit to store events...
```

Comment:
Presumably we want to aggregate string output events in memory rather than storing each individual one?